README.md
# Fast and flexible HTTP router for Go

[![build workflow](https://github.com/aikchun/bunrouter/actions/workflows/build.yml/badge.svg)](https://github.com/aikchun/bunrouter/actions)
[![PkgGoDev](https://pkg.go.dev/badge/github.com/aikchun/bunrouter)](https://pkg.go.dev/github.com/aikchun/bunrouter)
[![Documentation](https://img.shields.io/badge/bunrouter-documentation-informational)](https://bunrouter.uptrace.dev/)
[![Chat](https://discordapp.com/api/guilds/752070105847955518/widget.png)](https://discord.gg/rWtp5Aj)

> BunRouter is brought to you by :star: [**uptrace/uptrace**](https://github.com/uptrace/uptrace).
> Uptrace is an open-source APM tool that supports distributed tracing, metrics, and logs. You can
> use it to monitor applications and set up automatic alerts to receive notifications via email,
> Slack, Telegram, and others. Star it as well!

**TLDR** BunRouter is as fast as httprouter, but supports middlewares, routing rules priority, and
error handling.

BunRouter is an extremely fast HTTP router for Go with unique combination of features:

- [Middlewares](https://bunrouter.uptrace.dev/guide/golang-http-middlewares.html) allow to extract
  common operations from HTTP handlers into reusable functions.
- [Error handling](https://bunrouter.uptrace.dev/guide/golang-http-error-handling.html) allows to
  further reduce the size of HTTP handlers by handling errors in middlewares.
- [Routes priority](https://bunrouter.uptrace.dev/guide/golang-router.html#routes-priority) enables
  meaningful matching priority for routing rules: first static nodes, then named nodes, lastly
  wildcard nodes.
- net/http compatible API which means using minimal API without constructing huge wrappers that try
  to do everything: from serving static files to XML generation (for example, `gin.Context` or
  `echo.Context`).

| Router          | Middlewares        | Error handling     | Routes priority    | net/http API       |
| --------------- | ------------------ | ------------------ | ------------------ | ------------------ |
| BunRouter       | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| [httprouter][1] | :x:                | :x:                | :x:                | :heavy_check_mark: |
| [Chi][2]        | :heavy_check_mark: | :x:                | :heavy_check_mark: | :heavy_check_mark: |
| [Echo][3]       | :heavy_check_mark: | :heavy_check_mark: | :x:                | :x:                |
| [Gin][4]        | :heavy_check_mark: | :heavy_check_mark: | :x:                | :x:                |

[1]: https://github.com/julienschmidt/httprouter
[2]: https://github.com/go-chi/chi
[3]: https://github.com/labstack/echo
[4]: https://github.com/go-gin/gin

Learn:

- [Documentation](https://bunrouter.uptrace.dev/)
- [Reference](https://pkg.go.dev/github.com/aikchun/bunrouter)

Examples:

- [Basic example](/example/basic/)
- [http.HandlerFunc example](/example/basic-compat/)
- [httprouter.Handle example](/example/basic-verbose/)
- [CORS example](/example/cors/)
- [Basic auth example](/example/basicauth/)

Projects using BunRouter:

- [Distributed tracing tool](https://github.com/uptrace/uptrace)
- [input-output-hk/cicero](https://github.com/input-output-hk/cicero)
- [RealWorld example application](https://github.com/go-bun/bun-realworld-app)

Benchmarks:

- [web-frameworks-benchmark](https://web-frameworks-benchmark.netlify.app/result?l=go)
- [go-http-routing-benchmark](https://github.com/go-bun/go-http-routing-benchmark)

<details>
  <summary>Benchmark results</summary>

```
BenchmarkGin_Param               	16019718	        74.16 ns/op	       0 B/op	       0 allocs/op
BenchmarkHttpRouter_Param        	12560001	        95.04 ns/op	      32 B/op	       1 allocs/op
BenchmarkBunrouter_Param         	50015306	        23.81 ns/op	       0 B/op	       0 allocs/op
BenchmarkGin_Param5              	 8997234	       131.5 ns/op	       0 B/op	       0 allocs/op
BenchmarkHttpRouter_Param5       	 4809441	       261.3 ns/op	     160 B/op	       1 allocs/op
BenchmarkBunrouter_Param5        	10789635	       114.0 ns/op	       0 B/op	       0 allocs/op
BenchmarkGin_Param20             	 3953041	       302.4 ns/op	       0 B/op	       0 allocs/op
BenchmarkHttpRouter_Param20      	 1661373	       743.3 ns/op	     640 B/op	       1 allocs/op
BenchmarkBunrouter_Param20       	 2462354	       482.8 ns/op	       0 B/op	       0 allocs/op
BenchmarkGin_ParamWrite          	 9258986	       128.0 ns/op	       0 B/op	       0 allocs/op
BenchmarkHttpRouter_ParamWrite   	 9908178	       123.0 ns/op	      32 B/op	       1 allocs/op
BenchmarkBunrouter_ParamWrite    	15511226	        70.62 ns/op	       0 B/op	       0 allocs/op
BenchmarkGin_GithubStatic        	12781513	        94.17 ns/op	       0 B/op	       0 allocs/op
BenchmarkHttpRouter_GithubStatic 	30077443	        37.36 ns/op	       0 B/op	       0 allocs/op
BenchmarkBunrouter_GithubStatic  	37160334	        32.41 ns/op	       0 B/op	       0 allocs/op
BenchmarkGin_GithubParam         	 6971791	       169.2 ns/op	       0 B/op	       0 allocs/op
BenchmarkHttpRouter_GithubParam  	 5464755	       217.4 ns/op	      96 B/op	       1 allocs/op
BenchmarkBunrouter_GithubParam   	12047902	       101.2 ns/op	       0 B/op	       0 allocs/op
BenchmarkGin_GithubAll           	   32758	     37382 ns/op	       0 B/op	       0 allocs/op
BenchmarkHttpRouter_GithubAll    	   27324	     43932 ns/op	   13792 B/op	     167 allocs/op
BenchmarkBunrouter_GithubAll     	   57910	     20914 ns/op	       0 B/op	       0 allocs/op
BenchmarkGin_GPlusStatic         	17788194	        69.13 ns/op	       0 B/op	       0 allocs/op
BenchmarkHttpRouter_GPlusStatic  	60191341	        19.84 ns/op	       0 B/op	       0 allocs/op
BenchmarkBunrouter_GPlusStatic   	87114368	        14.06 ns/op	       0 B/op	       0 allocs/op
BenchmarkGin_GPlusParam          	10075399	       119.5 ns/op	       0 B/op	       0 allocs/op
BenchmarkHttpRouter_GPlusParam   	 8272046	       149.2 ns/op	      64 B/op	       1 allocs/op
BenchmarkBunrouter_GPlusParam    	37359979	        32.43 ns/op	       0 B/op	       0 allocs/op
BenchmarkGin_GPlus2Params        	 7375279	       162.9 ns/op	       0 B/op	       0 allocs/op
BenchmarkHttpRouter_GPlus2Params 	 6538942	       186.7 ns/op	      64 B/op	       1 allocs/op
BenchmarkBunrouter_GPlus2Params  	19681939	        61.51 ns/op	       0 B/op	       0 allocs/op
BenchmarkGin_GPlusAll            	  647716	      1752 ns/op	       0 B/op	       0 allocs/op
BenchmarkHttpRouter_GPlusAll     	  590356	      2085 ns/op	     640 B/op	      11 allocs/op
BenchmarkBunrouter_GPlusAll      	 1685287	       712.8 ns/op	       0 B/op	       0 allocs/op
BenchmarkGin_ParseStatic         	14566458	        76.58 ns/op	       0 B/op	       0 allocs/op
BenchmarkHttpRouter_ParseStatic  	52994076	        21.02 ns/op	       0 B/op	       0 allocs/op
BenchmarkBunrouter_ParseStatic   	50583933	        23.83 ns/op	       0 B/op	       0 allocs/op
BenchmarkGin_ParseParam          	13443874	        90.66 ns/op	       0 B/op	       0 allocs/op
BenchmarkHttpRouter_ParseParam   	 8825664	       135.6 ns/op	      64 B/op	       1 allocs/op
BenchmarkBunrouter_ParseParam    	38058278	        31.33 ns/op	       0 B/op	       0 allocs/op
BenchmarkGin_Parse2Params        	10179813	       118.1 ns/op	       0 B/op	       0 allocs/op
BenchmarkHttpRouter_Parse2Params 	 7801735	       152.9 ns/op	      64 B/op	       1 allocs/op
BenchmarkBunrouter_Parse2Params  	23704574	        50.78 ns/op	       0 B/op	       0 allocs/op
BenchmarkGin_ParseAll            	  394884	      3073 ns/op	       0 B/op	       0 allocs/op
BenchmarkHttpRouter_ParseAll     	  410238	      3011 ns/op	     640 B/op	      16 allocs/op
BenchmarkBunrouter_ParseAll      	  810908	      1487 ns/op	       0 B/op	       0 allocs/op
BenchmarkGin_StaticAll           	   50658	     23699 ns/op	       0 B/op	       0 allocs/op
BenchmarkHttpRouter_StaticAll    	  105313	     11518 ns/op	       0 B/op	       0 allocs/op
BenchmarkBunrouter_StaticAll     	   99674	     12188 ns/op	       0 B/op	       0 allocs/op
```

</details>

## Quickstart

Install:

```shell
go get github.com/aikchun/bunrouter
```

Run the [example](/example/basic/):

```go
package main

import (
	"html/template"
	"log"
	"net/http"

	"github.com/aikchun/bunrouter"
	"github.com/aikchun/bunrouter/extra/reqlog"
)

func main() {
	router := bunrouter.New(
		bunrouter.Use(reqlog.NewMiddleware()),
	)

	router.GET("/", indexHandler)

	router.WithGroup("/api", func(g *bunrouter.Group) {
		g.GET("/users/:id", debugHandler)
		g.GET("/users/current", debugHandler)
		g.GET("/users/*path", debugHandler)
	})

	log.Println("listening on http://localhost:9999")
	log.Println(http.ListenAndServe(":9999", router))
}

func indexHandler(w http.ResponseWriter, req bunrouter.Request) error {
	return indexTemplate().Execute(w, nil)
}

func debugHandler(w http.ResponseWriter, req bunrouter.Request) error {
	return bunrouter.JSON(w, bunrouter.H{
		"route":  req.Route(),
		"params": req.Params().Map(),
	})
}

var indexTmpl = `
<html>
  <h1>Welcome</h1>
  <ul>
    <li><a href="/api/users/123">/api/users/123</a></li>
    <li><a href="/api/users/current">/api/users/current</a></li>
    <li><a href="/api/users/foo/bar">/api/users/foo/bar</a></li>
  </ul>
</html>
`

func indexTemplate() *template.Template {
	return template.Must(template.New("index").Parse(indexTmpl))
}
```

See the [Golang Router documentation](https://bunrouter.uptrace.dev/) for details.

## See also

- [Golang ORM](https://bun.uptrace.dev/)
- [Golang msgpack](https://msgpack.uptrace.dev/)
- [Golang message task queue](https://github.com/vmihailenco/taskq)
- [Distributed tracing tools ](https://get.uptrace.dev/compare/distributed-tracing-tools.html)


CHANGELOG.md
## [1.0.20](https://github.com/aikchun/bunrouter/compare/v1.0.19...v1.0.20) (2023-02-12)



## [1.0.19](https://github.com/aikchun/bunrouter/compare/v1.0.18...v1.0.19) (2022-08-30)



## [1.0.18](https://github.com/aikchun/bunrouter/compare/v1.0.17...v1.0.18) (2022-08-09)


### Bug Fixes

* panic when trying to register same routes with different param names ([ace1f17](https://github.com/aikchun/bunrouter/commit/ace1f177c3bae9819612209e6fbb9c0496c8f58a))



## [1.0.17](https://github.com/aikchun/bunrouter/compare/v1.0.16...v1.0.17) (2022-05-25)


### Bug Fixes

* fix another corner case with incorrect param resolving ([7e3d720](https://github.com/aikchun/bunrouter/commit/7e3d7205a85cb90de5b3ddee3e2737e699fe810a))



## [1.0.16](https://github.com/aikchun/bunrouter/compare/v1.0.15...v1.0.16) (2022-05-23)


### Bug Fixes

* fix extracting params from routes with wildcard and named params ([2741aad](https://github.com/aikchun/bunrouter/commit/2741aadef5f5fdf135ffa541cf8b3b8644898ae0))


### Features

* **reqlog:** add trace id ([b2d6582](https://github.com/aikchun/bunrouter/commit/b2d658268dc2d573f5e9e9010a415c6102faa1d1))



## [1.0.15](https://github.com/aikchun/bunrouter/compare/v1.0.14...v1.0.15) (2022-05-08)


### Bug Fixes

* call method not allowed handler once ([37425b7](https://github.com/aikchun/bunrouter/commit/37425b7af9f103a7f26c7f8cc9a04ddac8b805b4))



## [1.0.14](https://github.com/aikchun/bunrouter/compare/v1.0.13...v1.0.14) (2022-04-14)


### Bug Fixes

* use httpsnoop to wrap response writer ([3d86932](https://github.com/aikchun/bunrouter/commit/3d869325af1302c55fdeb8a8ff7e4b7cfd10ece0))



## [1.0.13](https://github.com/aikchun/bunrouter/compare/v1.0.12...v1.0.13) (2022-03-22)


### Bug Fixes

* dedup middlewares applied on the method not allowed handler ([2659f03](https://github.com/aikchun/bunrouter/commit/2659f039be323b8bc4901e35f20a5760e6445640))


### Features

* accept multiple middlewares in Use(fn1, fn2, fn3) ([1486279](https://github.com/aikchun/bunrouter/commit/14862790b51c418384700a5e86aa49833937a772))



## [1.0.12](https://github.com/aikchun/bunrouter/compare/v1.0.11...v1.0.12) (2022-01-19)


### Bug Fixes

* use RawPath when available ([9859bc7](https://github.com/aikchun/bunrouter/commit/9859bc722310e0af24cd8372f585318379ccbbd9))



## [1.0.11](https://github.com/aikchun/bunrouter/compare/v1.0.10...v1.0.11) (2022-01-18)

### Features

- add basicauth middleware
  ([363da1a](https://github.com/aikchun/bunrouter/commit/363da1a989d943c8bbcf7551ad1a06150f6d1f1f))

* Added `Use` function which is an alias for `WithMiddleware`.

* Updated docs to use cleaner version of the API. Instead of:

```go
router.NewGroup("/some/prefix",
	 bunrouter.WithMiddleware(middleware1),
	 bunrouter.WithMiddleware(middleware2),
	 bunrouter.WithGroup(func(group *bunrouter.Group) {}),
)
```

You can use:

```go
router.Use(middleware1).
    Use(middleware2).
    WithGroup("/some/prefix", func(group *bunrouter.Group) {})
```

## [1.0.10](https://github.com/aikchun/bunrouter/compare/v1.0.9...v1.0.10) (2022-01-08)

- harden redir checks

## [1.0.9](https://github.com/aikchun/bunrouter/compare/v1.0.8...v1.0.9) (2021-12-23)

### Bug Fixes

- properly handle wildcard node without a slash
  ([88b4d3e](https://github.com/aikchun/bunrouter/commit/88b4d3ea352c92fc7a87972fc95add8e7f99c328))

### Features

- add Router.ServeHTTPError that returns the error from the handler
  ([9add167](https://github.com/aikchun/bunrouter/commit/9add167b91c37b42846a486a9965f5212d49bafa))

## [1.0.8](https://github.com/aikchun/bunrouter/compare/v1.0.7...v1.0.8) (2021-11-16)

### Bug Fixes

- properly handle empty root node
  ([b37ad45](https://github.com/aikchun/bunrouter/commit/b37ad4595c66454f4a768356298c95976e01d7f2))

## [1.0.7](https://github.com/aikchun/bunrouter/compare/v1.0.6...v1.0.7) (2021-11-12)

### Bug Fixes

- don't panic on path that matches routes common prefix
  ([d89dc38](https://github.com/aikchun/bunrouter/commit/d89dc38defc44bdf4bab13ecb518c2aa42ad9e80))

## [1.0.6](https://github.com/aikchun/bunrouter/compare/v1.0.5...v1.0.6) (2021-11-09)

### Bug Fixes

- propagate error in HTTP compat handlers
  ([5ed4d41](https://github.com/aikchun/bunrouter/commit/5ed4d41e99e8f6614753393f13e3674df29e7fb9))

## [1.0.5](https://github.com/aikchun/bunrouter/compare/v1.0.4...v1.0.5) (2021-11-08)

### Bug Fixes

- fallback to context when params can't be found
  ([ee2eb33](https://github.com/aikchun/bunrouter/commit/ee2eb3339ff421dd80566802304a32265f6e28b1))

### Features

- apply middleware to method not allowed handler
  ([8e295d0](https://github.com/aikchun/bunrouter/commit/8e295d0f01fbdf16061b7a4c53b931e9d709b25b))

## [1.0.4](https://github.com/aikchun/bunrouter/compare/v1.0.3...v1.0.4) (2021-11-07)

### Features

- **reqlog:** support http.Flusher
  ([938d70a](https://github.com/aikchun/bunrouter/commit/938d70aa4743d3c1492af8421a3fff14df986fa0))

## [1.0.3](https://github.com/aikchun/bunrouter/compare/v1.0.2...v1.0.3) (2021-10-21)

### Bug Fixes

- make routes with only colon nodes work
  ([fffd754](https://github.com/aikchun/bunrouter/commit/fffd75448f70a508254b0327c933cfda19eac70f))

## [1.0.2](https://github.com/aikchun/bunrouter/compare/v1.0.1...v1.0.2) (2021-10-19)

### Features

- make redirects work for wildcard routes
  ([04cb9f3](https://github.com/aikchun/bunrouter/commit/04cb9f3fd564d76477dcba7218e29f980503b15d))

## [1.0.1](https://github.com/aikchun/bunrouter/compare/v1.0.0...v1.0.1) (2021-10-15)

### Bug Fixes

- change WithContext to preserve route params
  ([2ca195a](https://github.com/aikchun/bunrouter/commit/2ca195ac8e7d9242d5110b84ede8d50a360f9a47))

# [1.0.0](https://github.com/aikchun/bunrouter/compare/v1.0.0-rc.2...v1.0.0) (2021-10-14)

### Bug Fixes

- make Slice and Map work on empty request
  ([609c7a3](https://github.com/aikchun/bunrouter/commit/609c7a3fcb6f5140c1def406efeee01eb0d80a11))

### Features

- allow configuring reqlog from env variables

example/rate-limiting/main.go
package main

import (
	"errors"
	"log"
	"net/http"
	"strconv"
	"time"

	"github.com/go-redis/redis/v8"
	"github.com/go-redis/redis_rate/v9"

	"github.com/aikchun/bunrouter"
	"github.com/aikchun/bunrouter/extra/reqlog"
)

var ErrRateLimited = errors.New("rate limited")

var limiter *redis_rate.Limiter

func main() {
	rdb := redis.NewClient(&redis.Options{
		Addr: "localhost:6379",
	})
	limiter = redis_rate.NewLimiter(rdb)

	router := bunrouter.New(
		bunrouter.Use(reqlog.NewMiddleware()),
		bunrouter.Use(errorHandler),
		bunrouter.Use(rateLimit),
	)

	router.GET("/", indexHandler)

	log.Println("listening on http://localhost:9999")
	log.Println(http.ListenAndServe(":9999", router))
}

func indexHandler(w http.ResponseWriter, req bunrouter.Request) error {
	_, err := w.Write([]byte("hello world"))
	return err
}

func rateLimit(next bunrouter.HandlerFunc) bunrouter.HandlerFunc {
	return func(w http.ResponseWriter, req bunrouter.Request) error {
		res, err := limiter.Allow(req.Context(), "project:123", redis_rate.PerMinute(10))
		if err != nil {
			return err
		}

		h := w.Header()
		h.Set("RateLimit-Remaining", strconv.Itoa(res.Remaining))

		if res.Allowed == 0 {
			// We are rate limited.

			seconds := int(res.RetryAfter / time.Second)
			h.Set("RateLimit-RetryAfter", strconv.Itoa(seconds))

			// Stop processing and return the error.
			return ErrRateLimited
		}

		// Continue processing as normal.
		return next(w, req)
	}
}

func errorHandler(next bunrouter.HandlerFunc) bunrouter.HandlerFunc {
	return func(w http.ResponseWriter, req bunrouter.Request) error {
		err := next(w, req)

		switch err {
		case nil:
			// ok
		case ErrRateLimited:
			w.WriteHeader(http.StatusTooManyRequests)
			_ = bunrouter.JSON(w, bunrouter.H{
				"message": "you are rate limited",
			})
		default:
			w.WriteHeader(http.StatusInternalServerError)
			_ = bunrouter.JSON(w, bunrouter.H{
				"message": err.Error(),
			})
		}

		return err
	}
}
